File: Reqnroll/Chapter20/20.2-TestConfigurationProviderPattern/Before/WIMP.Specs/Support/LambdaAction.cs
namespace WIMP.Specs.Support;

public class LambdaAction<TResult>(string testActionName, Func<Task<TResult>> action)
    : TestAction<TResult>(testActionName)
{
    /// <summary>
    /// Creates a lambda action with a synchronous action
    /// </summary>
    public LambdaAction(string testActionName, Func<TResult> action)
        : this(testActionName, () => Task.FromResult(action()))
    {
    }

    protected override async Task<TResult> DoExecute()
    {
        return await action();
    }
}

/// <summary>
/// Lambda action for void-return actions
/// </summary>
public class LambdaAction(string testActionName, Func<Task> action) :
    LambdaAction<VoidReturn>(testActionName, async () =>
    {
        await action();
        return VoidReturn.Instance;
    })
{
    /// <summary>
    /// Creates a void-return lambda action with a synchronous action
    /// </summary>
    public LambdaAction(string testActionName, Action action)
        : this(testActionName, () =>
        {
            action();
            return Task.CompletedTask;
        })
    {
    }
}

File: CucumberJVM/Chapter19/19.2-SupplementaryAttachmentPattern/Before/src/test/java/com/wimp/app/specs/support/LambdaAction.java
package com.wimp.app.specs.support;

public class LambdaAction<T> extends TestAction<T> {

    /**
     * Helper interface for specifying lambda actions that throw exceptions.
     */
    @FunctionalInterface
    public interface ThrowingSupplier<T> {
        T get() throws Exception;
    }

    private final ThrowingSupplier<T> action;

    public LambdaAction(String testActionName, Object input, ThrowingSupplier<T> action) {
        super(testActionName, input);
        this.action = action;
    }

    public LambdaAction(String testActionName, ThrowingSupplier<T> action) {
        this(testActionName, null, action);
    }

    @Override
    public T doExecute() throws Exception {
        return action.get();
    }

    /**
     * Lambda action for void-returning actions.
     */
    public static class Void extends LambdaAction<VoidReturn> {

        @FunctionalInterface
        public interface ThrowingRunnable {
            void run() throws Exception;
        }

        public Void(String testActionName, Object input, ThrowingRunnable action) {
            super(testActionName, input, () -> {
                action.run();
                return VoidReturn.INSTANCE;
            });
        }

        public Void(String testActionName, ThrowingRunnable action) {
            this(testActionName, null, action);
        }
    }
}
